Template file for gluster-s3 container

SaravanaStorageNetwork opened the pull request:
Add a template file for gluster-s3 container.
Add corresponding README having worflow and testing instructions.

Signed-off-by: Saravanakumar <sarumuga@redhat.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gluster/gluster-kubernetes/277)
<!-- Reviewable:end -->

sabose commented on deploy/ocp-templates/glusters3template.json:
```diff
@@ -0,0 +1,205 @@
+        "creationTimestamp": null,
+        "annotations": {
+          "description": "Gluster s3 service template",
+          "tags": "glusterfs,heketi"
```
Do we need heketi tag here? maybe gluster-s3 as tag?

SaravanaStorageNetwork replied:
Actually we depend upon heketi to create volume..although internally. will add gluster-s3 tag as well.

sabose commented on deploy/ocp-templates/glusters3template.json:
```diff
@@ -0,0 +1,205 @@
+                "containers": [
+                    {
+                        "name": "glusters3",
+                        "image": "brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhgs3/rhgs-s3-server-rhel7:3.3.0-2",
```
The registry path should not be hardcoded? Won't  rhgs3/rhgs-s3-server-rhel7:3.3.0-2 suffice?

SaravanaStorageNetwork replied:
Done

SaravanaStorageNetwork replied:
@jarrpa Done

raghavendra-talur commented on deploy/ocp-templates/storageclass.yaml:
```diff
@@ -0,0 +1,9 @@
+apiVersion: storage.k8s.io/v1beta1
```
I think this should go to examples and not templates.

SaravanaStorageNetwork replied:
Ok, moving to examples/gluster-s3-storage-template/

raghavendra-talur commented on deploy/ocp-templates/glusters3template.json:
```diff
@@ -0,0 +1,205 @@
+{
+    "kind": "Template",
```
Can we have the template in yaml please..All the other templates have been ported to yaml for ocp-templates

SaravanaStorageNetwork replied:
Done

jarrpa replied:
Also move this file into the examples directory. It should not be in `deploy/ocp-templates` unless/until it is used by a script.

SaravanaStorageNetwork replied:
@jarrpa  ok moving to examples directory

sabose commented on deploy/ocp-templates/glusters3template.json:
```diff
@@ -0,0 +1,205 @@
+                "name": "glusterfs-claim",
+                "creationTimestamp": null,
+                "annotations": {
+                    "volume.beta.kubernetes.io/storage-class": "s3storageclass"
```
Name of storage class should be a parameter, with default value of s3storageclass perhaps?

SaravanaStorageNetwork replied:
Done..adding a param.

sabose commented on deploy/ocp-templates/glusters3template.json:
```diff
@@ -0,0 +1,205 @@
+                "name": "glusterfs-claim-meta",
+                "creationTimestamp": null,
+                "annotations": {
+                    "volume.beta.kubernetes.io/storage-class": "s3storageclass"
```
same as above

SaravanaStorageNetwork replied:
Done..adding a param.

sabose commented on deploy/ocp-templates/glusters3template.json:
```diff
@@ -0,0 +1,205 @@
+    ],
+  "parameters": [
+    {
+      "name": "GLUSTER_VOLUMES",
```
If the gluster volume is created dynamically, why would this name be needed?

SaravanaStorageNetwork replied:
We need a string to be passed to gluster-s3 container for swift-genbuilders and for mounting at specific location.

sabose replied:
I think the description should be updated accordingly. It's misleading at the moment

SaravanaStorageNetwork replied:
Agreed.. this is bit confusing. 

Actually it is a s3 storage account name with underlying Gluster volume which gets mounted at /mnt/gluster-object/<account name> . I will update the description.

sabose commented on deploy/ocp-templates/glusters3template.json:
```diff
@@ -0,0 +1,205 @@
+    {
+      "name": "VOLUME_CAPACITY",
+      "displayName": "Volume capacity",
+      "description": "Volume space available for data, e.g. 1Gi, 2Gi.",
```
Capacity of gluster object store ?

SaravanaStorageNetwork replied:
Done

SaravanaStorageNetwork commented:
@jarrpa  rename done

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,109 @@
+
+# Set glusters3 service using template:
+
+oc new-app gluster-s3-template.yaml  --param=GLUSTER_VOLUMES=testvolume  --param=GLUSTER_USER=adminuser --param=GLUSTER_PASSWORD=itsmine --param=VOLUME_CAPACITY=2Gi
```
Surround this line with ``` at the beginning and end.

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,109 @@
+Note: adjust parameters according to your needs.
+
+Available at:
+docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml
```
Change this to `[gluster-s3-template.yaml](./gluster-s3-template.yaml)`

SaravanaStorageNetwork replied:
Done

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,109 @@
+docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml
+
+For example:
+===========================
```
Remove this line, insert blank line

SaravanaStorageNetwork replied:
Done

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,109 @@
+    persistentvolumeclaim "glusterfs-claim" created
+    persistentvolumeclaim "glusterfs-claim-meta" created
+--> Success
+    Run 'oc status' to view your app.
```
Surround everything from the `oc new-app` command to here with `` ``` `` at the beginning and end

SaravanaStorageNetwork replied:
Done

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,109 @@
+    persistentvolumeclaim "glusterfs-claim-meta" created
+--> Success
+    Run 'oc status' to view your app.
+------------------------------
```
Remove this line, replace with blank line

SaravanaStorageNetwork replied:
Done

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,109 @@
+glusterfs-85qds                  1/1       Running   0          4d        10.70.42.5     node1
+glusters3                        1/1       Running   0          4m        10.130.0.29    node3
+heketi-1-m8817                   1/1       Running   0          4d        10.130.0.19    node3
+storage-project-router-1-2816m   1/1       Running   0          4d        10.70.42.234   node3
```
Surorund this block with `` ``` `` at the beginning and end

SaravanaStorageNetwork replied:
Done

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,109 @@
+heketi-1-m8817                   1/1       Running   0          4d        10.130.0.19    node3
+storage-project-router-1-2816m   1/1       Running   0          4d        10.70.42.234   node3
+
+==============================
```
Remove this line

SaravanaStorageNetwork replied:
Done

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,109 @@
+
+==============================
+
+Testing:
```
Add `#` at the beginning, remove `:`

SaravanaStorageNetwork replied:
Done

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,109 @@
+# Get IP addreess of glusters3 service
+ip_address=$(oc get svc   | grep glusters3   | awk '{print $2}')
+
+---------------------------
```
Remove this line

SaravanaStorageNetwork replied:
Done

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,109 @@
+# Verify bucket Delete request
+s3curl.pl --debug --id "testvolume:adminuser" --key "itsmine" --delete  --  http://$ip_address:8080/bucket1
+
+---------------------------
```
Remove the above two lines

SaravanaStorageNetwork replied:
Done

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,109 @@
+
+
+This step should not be required, once the delay is updated in the image.
+=================================================
```
Remove this line

SaravanaStorageNetwork replied:
I will remove the entire note section..it no longer required with new build.

SaravanaStorageNetwork commented:
<<You don't mention why you want to get the ip_address variable. 
updated with explanation.
<<The my @endpoints line is not proper shell script, I can only assume this belongs somewhere in Perl-land
yes it is a perl script..updating same.
<<You don't explain what s3curl.pl is or where to get it
Added now.
<<The Note at the end seems unrelated to everything above, or at least doesn't explain how it's related nor when it should be run.
Removing it - it is a workaroud should no longer be required with new build.

SaravanaStorageNetwork commented on docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml:
```diff
@@ -0,0 +1,134 @@
+- kind: Service
+  apiVersion: v1
+  metadata:
+    name: glusters3service
```
route is added now check for  glusters3object in template file.

sabose replied:
Great, thanks!

SaravanaStorageNetwork commented on docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml:
```diff
@@ -0,0 +1,134 @@
+- kind: PersistentVolumeClaim
+  apiVersion: v1
+  metadata:
+    name: glusterfs-claim
```
Done

SaravanaStorageNetwork commented on docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml:
```diff
@@ -0,0 +1,134 @@
+  required: true
+- name: STORAGE_CLASS
+  displayName: Storage class
+  description: Storage class for creating underlying Volume
```
Done

sabose commented on docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml:
```diff
@@ -133,8 +133,8 @@ parameters:
-  displayName: Storage class
-  description: Storage class for creating underlying Volume
+  displayName: Storage class with GlusterFS provisioner
+  description: Storage class with GlusterFS provisioner for creating gluster volumes
```
Thanks!

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,202 @@
+If you wish to make use of any existing storage class, add another parameter like
+
+
+--param=STORAGE_CLASS=<your storage class name>
```
Surround this line with `` ``` ``.

SaravanaStorageNetwork replied:
Done

jarrpa commented on docs/examples/gluster-s3-storage-template/README.md:
```diff
@@ -0,0 +1,202 @@
+# Gluster S3 docker container using template
```
Change this to something like `Gluster S3 (gluster-object) as native app on OpenShift`.

SaravanaStorageNetwork replied:
Done..will reword it as - Gluster S3 object storage as native app on OpenShift.

SaravanaStorageNetwork commented:
<<If it makes sense, could you provide sample output for all the other s3curl.pl commands?
It makes too verbose..there is not much user get out of this.

<<Also, I'd love to see a version of this for Kubernetes, but that can come as a later enhancement. :)
Sure

SaravanaStorageNetwork commented:
@jarrpa  I have addressed the review comments. please check.

humblec commented:
@jarrpa Can you please review this at the earliest?

humblec commented:
@SaravanaStorageNetwork just for sanity, we could test s3curl from app pod and verify the service is accessible from app pod.

SaravanaStorageNetwork commented:
@humblec  I have updated to use "oc create template"

SaravanaStorageNetwork commented:
@humblec  @jarrpa   Please review and merge if no other comment

jarrpa commented:
@humblec @SaravanaStorageNetwork I believe `oc new-app` should be fine, so please remove the latest commit. Out of curiosity, does the output of `oc get template --all-namespaces` indicate that the actual template gets created when you use `new-app`?

SaravanaStorageNetwork commented:
@jarrpa  When I use oc new-app, I don't see templates created in "oc get template --all-namespaces".
From this, I infer, (by using oc new-app) all objects in the template file is launched but template is NOT created.

I think it makes sense to use "oc create" followed by "oc process" as templates are visible and can be used to launch s3 service further.

Let me know if you still insist on removing last commit.

jarrpa commented:
@SaravanaStorageNetwork Hmm... well, I don't immediately see a need for it, especially in the case of an example. Further, the way the template is structured right now makes it difficult to reuse given that it would try to create new resources under the same names. Thus, either update the template to make it truly reusable or revert to the `new-app` method.

SaravanaStorageNetwork commented:
@jarrpa  agreed.. will revert to use oc new-app for now.
